compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala

```diff
@@ -0,0 +1,147 @@
+package dotty.tools.dotc.core
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Symbols.defn
+import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Type, TypeBounds, TypeMap, TypeProxy}
+
+/** Defines operations on nullable types. */
+object NullOpsDecorator {
+
+  implicit class NullOps(val self: Type) {
```

Comment:
I am a bit concerned about efficiency of everything here. This could become a major bottleneck. Everywhere else in the compiler we work hard to avoid duplicated computations and object allocations as much as possible. We should do the same here.

Reply:
Done

Reply:
We have optimized things in this file and avoided allocations when possible.